pbxd/pbx/definity.py

```diff
@@ -0,0 +1,446 @@
+            elif index == 1:  # EOF
+                response_error = 'PBX connection failed with EOF'
+                self.logger.error('{}: {}\n{}'.format(response_error, command, self.session.before))
+            elif index >= 2:
+                self.logger.debug('{} saving screen:\n{}'.format(index, self.session.after))
+                screen = pyte.Screen(80, 24)
+                stream = pyte.Stream(screen)
+                stream.feed(self.session.before.decode('utf-8'))
+                text = '\n'.join(screen.display)
+                screens.append(text)
+
+                if index == 2:  # check for an error message at the command prompt
+                    pbx_message = screen.display[22].strip()
+                    if pbx_message != '' and pbx_message != 'Command successfully completed':
+                        response_error = screen.display[22].strip()  # error is on line 23
+                        self.logger.warning(response_error)
+
+                elif index == 3:  # press NEXT PAGE to continue
+                    more_pages = True
+                    self.session.send(b'\x1b[6~')  # VT220 next page
+
+                else:  # look for paging
+                    m = re.match(r'.*Page +(\d+) of +(\d+).*', self.session.before.decode('utf-8'))
+                    if m is not None and m.group(1) < m.group(2):
+                        self.logger.debug('page {} of {}: requesting next page'.format(m.group(1), m.group(2)))
+                        more_pages = True
+                        self.session.send(b'\x1b[6~')  # VT220 next page
+
+        self.logger.info('command complete')
+
+        # return to the vt220 prompt and consume it
+        self.session.send(b'\x1b[3~')  # VT220 cancel
+        index = self.session.expect(
+            [
+                pexpect.TIMEOUT,
+                pexpect.EOF,
+                r'\[KCommand:',
+            ]
+        )
+        if index == 0:  # TIMEOUT
+            response_error = 'Timeout on vt220_command'
+            self.logger.error('{}: {}\n{}'.format(response_error, command, self.session.before))
+        elif index == 1:  # EOF
+            response_error = 'Connection failed with EOF on vt220_command'
+            self.logger.error('{}: {}\n{}'.format(response_error, command, self.session.before))
+
+        response_obj = {"screens": screens}
+        if response_error is not None:
+            response_obj['error'] = response_error
+        return response_obj
+
+    def send_pbx_command(self, termtype, command, fields, debug=False):
+        """
+        run a command with the requested termtype
+        """
+
+        if termtype == self.Termtype.vt220.name:
+            return self.vt220_command(command)
+        elif termtype == self.Termtype.ossi.name:
+            return self.ossi_command(command, fields=fields, debug=debug)
+        else:
+            return {"error": "Unknown termtype. Must be ossi or vt220."}
```

Comment:
I think this should die instead of returning this error structure. Have the web API import the `Termtype` enum and validate the input before calling `send_pbx_command`. This is some real nitpicking about division of responsibility at this point so I won't block on it. But conceptually the `definity` module shouldn't know anything about the webapp or the kind of error response format it expects.

Reply:
I see what you mean. I am going to keep it this way for a few reasons. Currently the views are very simple and only enforce of the structure of the data and not the content. This is a content error that needs to flow all the way back to the client just like if it came directly from the PBX (in theory it could if this supported any terminal type).

I reviewed the other places where the 'error' and Exceptions are used and they follow the same pattern where issues that the client can resolve in the command are reported in the 'error' element and returned, while issues that the system operator must resolve or investigate raise an exception and get logged.